### PR TITLE
[FTL-1525] Use testmail.app for OTP tests & additional support for message properties

### PR DESCRIPTION
**Pull request by vitalyrudenko-affinidi**
Changelog:
- Use [testmail.app](https://testmail.app/docs/) for all OTP integration tests in `core`, `browser`, `react-native` and `expo` packages
- Refactored `TestmailInbox` to use [livequery](https://testmail.app/docs/#waiting-for-new-email) (waiting for new email)
- Add support for `messageProperties` option in `CommonNetworkMember#changeUsername()`, `CommonNetworkMember#resendSignUpConfirmationCode()` and `CognitoService#changeUsername()`
- Fixed linter and type issues with `tiny-lds-ecdsa-secp256k1-2019` (updated version to `1.0.2`)

TODO for subsequent PRs:
- [ ] Don't use `{{CODE}}` in the **subject** of OTP emails for passwordless login's template since it's discouraged by AWS ([link 1](https://forums.aws.amazon.com/message.jspa?messageID=870189), [link 2](https://github.com/aws-amplify/amplify-js/issues/936)) and actually not supported by the default auth flows at all
- [ ] Update `create-auth-challenge` lambda to not to replace `{{CODE}}` in the **subject** of OTP emails

### CODE REVIEW

File: sdk/core/test/integration/otp/index.test.ts [hunk 1]
```diff
@@ -1,417 +1,426 @@
...
+  { suffix }: { suffix?: string } = {},
+): [string, string, MessageParameters] => {
+  const tag = generateTag(testId, suffix)
+  const email = TestmailHelper.generateEmailForTag(tag)
+  const messageParameters: MessageParameters = {
+    message: `Your verification code is: {{CODE}}`,
+    subject: `Verification code`,
+  }
+
+  return [email, tag, messageParameters]
```
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:31]: It's more convenient to use when I need new credentials for changing username, for example:
```ts
const [email, tag, messageParameters] = ...
...
const [newEmail, newTag, newMessageParameters] = ...
...
```

File: sdk/core/test/integration/otp/index.test.ts [hunk 2]
```diff
@@ -1,417 +1,426 @@
...
-    expect(networkMember).to.be.an.instanceof(CommonNetworkMember)
+  describe('[with existing user]', () => {
+    let email: string
+    let tag: string
+    let messageParameters: MessageParameters
+
+    // reuse the same user for all subsequent tests to reduce OTP emails count
+    before(async () => {
+      // testId is not available in "before" hook, so we generate it separately
+      ;[email, tag, messageParameters] = prepareOtpMessageParameters(generateTestId())
```
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:320]: Not sure. I also thought about it. Perhaps generating it once (where possible, like now) reduces the risk of collision even more.
It's not that important, though, since emails are only "waited" since specified timestamp.
Subjectively it just looks somewhat "cleaner" when we have an id for a test run instead of generating random string each time we need new email. And it also corresponds to the documentation of testmail.

File: sdk/core/test/integration/otp/index.test.ts [hunk 3]
```diff
@@ -1,417 +1,426 @@
...
 const options: SdkOptions = getOptionsForEnvironment()
+const { env } = getOptionsForEnvironment(true)
+
+const prepareOtpMessageParameters = (
+  testId: string,
+  { suffix }: { suffix?: string } = {},
+): [string, string, MessageParameters] => {
+  const tag = generateTag(testId, suffix)
+  const email = TestmailHelper.generateEmailForTag(tag)
+  const messageParameters: MessageParameters = {
```
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:26]: As long as it's not modified somewhere deep in the code, yeah, I agree.

File: sdk/core/test/integration/otp/index.test.ts [hunk 4]
```diff
@@ -1,417 +1,426 @@
...
 
 const options: SdkOptions = getOptionsForEnvironment()
+const { env } = getOptionsForEnvironment(true)
+
+const prepareOtpMessageParameters = (
+  testId: string,
+  { suffix }: { suffix?: string } = {},
+): [string, string, MessageParameters] => {
+  const tag = generateTag(testId, suffix)
+  const email = TestmailHelper.generateEmailForTag(tag)
```
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:25]: I actually thought about it, but since it doesn't change the complexity in the test themselves, I don't think it's better (or worse) to do it that way. I'd rather have these actions separated and have small helpers on top of the file. 
I don't like parsing things unnecessary, because it can make email generation rigid and also raise "wtf?" questions when we come back to the code later.

File: sdk/browser/package-lock.json [hunk 5]
```diff
@@ -83,6 +83,22 @@
 			"integrity": "sha512-Zywpj3D/su2oZOmMLgfaQc08Aqkwkm2O8BgHTMMgZ675KakFpTYtOZSxtrl3j/+XpA0g5NPuC486C51d3JrCGA==",
 			"dev": true
 		},
+		"@affinidi/issuer-email-ses-client": {
```
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/browser/package-lock.json]: I think this is the result of the `lerna clean && lerna bootstrap`

File: sdk/core/test/helpers/TestmailHelper.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+interface TestmailEmail {
+  subject: string
+  html?: string
+  text?: string
+}
+
+// NOTE: Don't use this helper for mere email generation,
+//       since there are limits for emails sent per month
+//       Use generateEmail() for random email generation
+export class TestmailHelper {
```
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/helpers/TestmailHelper.ts:18]: How would you name the file?

File: sdk/core/test/integration/otp/index.test.ts [hunk 7]
```diff
@@ -1,417 +1,426 @@
...
+  const messageParameters: MessageParameters = {
+    message: `Your verification code is: {{CODE}}`,
+    subject: `Verification code`,
+  }
+
+  return [email, tag, messageParameters]
+}
+
+const waitForOtpCode = async (tag: string, timestampFrom?: number): Promise<string> => {
+  const { text, html } = await TestmailHelper.waitForNewEmail(tag, timestampFrom)
```
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:35]: `livequery` immediately returns the result if there is at least one matching email in the box, so your proposed logic might look too complicated in the actual code
I think we should concern ourselves with such questions only if we have concrete problems to solve.
If clock is faulty, then there are more problems then just integration tests failing - which we will see immediately.

File: sdk/core/test/integration/otp/index.test.ts [hunk 8]
```diff
@@ -1,417 +1,426 @@
...
+      const timestamp = Date.now()
+      const token = await CommonNetworkMember.passwordlessLogin(email, options, messageParameters)
+      const otpCode = await waitForOtpCode(tag, timestamp)
+
+      const commonNetworkMember = await CommonNetworkMember.completeLoginChallenge(token, otpCode, options)
+      expect(commonNetworkMember).to.be.instanceOf(CommonNetworkMember)
+    })
+
+    // requires test timeout >= 180000 ms
+    it.skip('#completeLoginChallenge should throw "COR-17" for expired OTP', async () => {
```
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:372]: Sounds interesting, forgot about per-test configuration. Will try!

File: sdk/core/test/integration/otp/index.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (inga-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:320]: Then, since it is synchronous, maybe it is better to move it out of the top-level `describe` and just do `const testId = generateTestId()` at the top of the module? That way all tests will be able to use it; you won't have to think whether it is available in "before" hook or not.

File: sdk/core/test/helpers/TestmailHelper.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (inga-affinidi) [reply sdk/core/test/helpers/TestmailHelper.ts:18]: testMailHelpers, for example?

File: sdk/core/test/integration/otp/index.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:320]: Sorry for the confusion. The line we're currently discussing is an exception (since we reuse the user account).
"testId" should be unique in each spec. Though, now that I've said it, it would be easier to generate new id every time we generate an email.
I will remove it.

File: sdk/core/test/integration/otp/index.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (inga-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:35]: > livequery immediately returns the result if there is at least one matching email in the box

Which is why I said "call api with livequery=true **and timestamp=lastMessageTimestamp**" :)

> If clock is faulty, then there are more problems then just integration tests failing

It does not have to be faulty, it just has to be a couple of seconds behind or ahead of clock on their server. It happens all the time, that's why we have NTP to keep it in check :)
Generally, relying on the fact that your clock is in perfect sync with remote clock is quite a bad idea; it is not guaranteed to be.
And there is a way to avoid relying on it (by supplying testmail with timestamps they generate, instead of our timestamps).

File: sdk/core/test/integration/otp/index.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (inga-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:26]: It should not be modified by the code under testing. If it is, that's a bug that should be fixed :)

File: sdk/core/test/integration/otp/index.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:35]: Yeah, I agree, we can utilize their timestamps :+1:

File: sdk/core/test/integration/otp/index.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (inga-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:25]: > but since it doesn't change the complexity in the test themselves

But it does! Right now, tests have to concern themselves with implementation details of TestMailHelper; they have to know that in addition to email there is some "tag" used to access messages.

On the other hand, we could avoid parsing things by storing the tag in the `Mailbox` object, as proposed in other comment...

[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:25]: Oh, I meant the specs, not the helpers around it. In combination with your mailbox proposal, it makes sense to move all these helpers into one place. I will try to do that, thank you

File: sdk/core/test/integration/otp/index.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (inga-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:35]: (In the context of another discussion there) I would suggest to do something along the lines of:
```
class Mailbox {
  readonly email: string
  private readonly tag: string
  private lastMessageTimestamp = 0

  Mailbox() {
    this.tag = generateTag()
    this.email = generateEmailFromTag()
  }

  fetchNewMessage() {
    const url = new URL(TESTMAIL_API_URL)
    url.searchParams = new URLSearchParams({
      tag: this.tag,
      timestamp_from: this.lastMessageTimestamp,
      livequery: 'true',
      ...
    )
    const response = await fetch(url)
    const { emails } = await response.json()
    const { timestamp, subject, html, text } = emails[0]
    this.lastMessageTimestamp = timestamp
    return { subject, html, text }
  }
}
```

That way we would not have to parse email addresses for tags.

File: sdk/core/test/integration/otp/index.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (inga-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:25]: Isn't this the test spec? :)
https://github.com/affinityproject/affinidi-core-sdk/blob/012bcd9b27a0054030f4197512a63910d4521456/sdk/core/test/integration/otp/index.test.ts#L59-L68

[the remark this answers is not in the document]
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/integration/otp/index.test.ts:25]: Well I wanted to test "messageParameters" separately and then just use the "waitForOtpCode" method everywhere, but I see your point

File: sdk/core/src/services/CognitoService.ts [hunk 9]
```diff
@@ -278,11 +278,8 @@ export default class CognitoService {
 
     const { clientId: ClientId } = this.cognitoOptions
 
-    const params = { ClientId, Username }
-
-    if (messageParameters) {
-      Object.assign(params, { ClientMetadata: messageParameters })
-    }
+    const ClientMetadata = messageParameters || {}
+    const params = { ClientId, Username, ClientMetadata }
```
Comment (DenisPopov15) [inline sdk/core/src/services/CognitoService.ts:282]: @VitaliiShpital did you run all tests and their working as expected for all custom message use cases, correct ?
Wondering if nothing get broken when ClientMetadata present and empty (just thinking why it was written this way before)

File: sdk/core/test/integration/index.test.ts [hunk 10]
```diff
@@ -1,5 +1,6 @@
 import 'mocha'
-
 import './env'
+
 require('./services')
 require('./NetworkMember.test')
+require('./otp/index.test')
```
[the remark this answers is not in the document]
Comment (DenisPopov15) [reply sdk/core/test/integration/index.test.ts:6]: Lets have one entry point for the test suits, then number of features works as expected (as `.only`, etc). So I vote leave it this way as it done here

**Review by Volland — DISMISSED** [review]

File: sdk/browser/test/helpers/TestmailInbox.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+  private _email: string
+  private _lastEmailTimestamp: number
+
+  constructor({ prefix, suffix }: InboxOptions) {
+    const inboxId = cryptoRandomString({ length: 12 }) // arbitrary length
+    this._tag = `${prefix}.${inboxId}.${suffix}`
+    this._email = `${TESTMAIL_NAMESPACE}.${this._tag}@${TESTMAIL_INBOX_DOMAIN}`
+
+    // if inbox has already been used, ignore its old emails
+    this._lastEmailTimestamp = Date.now() - 60_000
```
Comment (inga-affinidi) [inline sdk/browser/test/helpers/TestmailInbox.ts:34]: One minute?

File: sdk/expo/test/helpers/TestmailInbox.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import fetch from 'node-fetch'
```
Comment (inga-affinidi) [inline sdk/expo/test/helpers/TestmailInbox.ts:1]: Looks like it's time to move out this file to some shared package instead of copy&pasting it into different projects.

What do you think, @DenisPopov15 ?

File: sdk/browser/test/helpers/TestmailInbox.ts [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+interface Email {
+  subject: string
+  body: string
+}
+
+// NOTE: Don't use this helper for mere email generation,
+//       since there are limits for emails sent per month
+//       Use generateEmail() instead
```
Comment (inga-affinidi) [inline sdk/browser/test/helpers/TestmailInbox.ts:22]: I do not understand what this comment means.
Does "email" in "for mere email generation" refer to the address or to the message?
When should I use this helper and when should I use `generateEmail`?
Other developers in the future might also have problems with understanding it.

File: sdk/browser/test/integration/otp/index.test.ts [hunk 14]
```diff
@@ -4,26 +4,17 @@ import '../env'
...
-
-const generateEmail = () => {
-  const TIMESTAMP = Date.now().toString(16).toUpperCase()
+import { getOptionsForEnvironment } from '../../helpers'
+import { TestmailInbox } from '../../../test/helpers/TestmailInbox'
+import { openAttestationDocument } from '../../factory/openAttestationDocument'
 
-  return `test.user-${TIMESTAMP}@gdwk.in`
-}
+const signedCredentials = require('../../factory/signedCredentials')
```
Comment (inga-affinidi) [inline sdk/browser/test/integration/otp/index.test.ts:14]: You can also use import there :)

File: sdk/browser/test/helpers/TestmailInbox.ts (hunk 11 above)
Comment (inga-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:34]: And why would we need it at all if inboxId is cryptographically secure random? If you think there are meaningful chances of collision, then we should probably increase the length of `inboxId` instead.

File: sdk/browser/test/integration/otp/index.test.ts [hunk 15]
```diff
@@ -39,174 +30,154 @@ const credentialShareRequestToken =
...
+const createInbox = () => new TestmailInbox({ prefix: env, suffix: 'otp.browser' })
 
-    const networkMember = await AffinityWallet.confirmSignUp(token, signUpOtp, options)
+describe('AffinityWallet [OTP]', () => {
+  it('Save Open Attestation credential and #deleteCredential scenario', async () => {
+    const inbox = createInbox()
+    const password = COGNITO_PASSWORD
 
-    let credentials
+    const signUpToken = await AffinityWallet.signUp(inbox.email, password, options, messageParameters)
```
Comment (inga-affinidi) [inline sdk/browser/test/integration/otp/index.test.ts:52]: By the way, since we're here: @DenisPopov15 , do we really want to use `TEST_SECRETS.COGNITO_PASSWORD` here instead of generating a new password every time? 🤔
(This code was there for a long time, so the comment is not related to this PR)

File: sdk/core/test/integration/index.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (inga-affinidi) [reply sdk/core/test/integration/index.test.ts:6]: @DenisPopov15 , I'm not sure what you mean? What is the problem with `.only`?

When there are multiple test files, and only one test in one of them is marked as `.only`, only that one test is run. (It already works this way in cloud-wallet-api)

File: sdk/core/test/unit/services/CognitoService.test.ts [hunk 16]
```diff
@@ -711,16 +712,35 @@ describe('CognitoService', () => {
   })
 
   describe('#changeUsername', () => {
-    it(successPathTestName, async () => {
-      const error = { code: USER_NOT_FOUND_EXCEPTION }
+    describe('[username is not taken]', () => {
+      beforeEach(() => {
+        stubMethod(INITIATE_AUTH, null, { code: USER_NOT_FOUND_EXCEPTION })
```
Comment (inga-affinidi) [inline sdk/core/test/unit/services/CognitoService.test.ts:717]: Shouldn't it be `beforeAll`?

File: sdk/expo/test/helpers/TestmailInbox.ts (hunk 12 above)
Comment (inga-affinidi) [reply sdk/expo/test/helpers/TestmailInbox.ts:1]: But also I don't see if this file is referenced anywhere in `expo`?

File: sdk/core/src/services/CognitoService.ts (hunk 9 above)
Comment (vitalyrudenko-affinidi) [reply sdk/core/src/services/CognitoService.ts:282]: Yeah nothing have changed with this tweak. I've actually copy-pasted this from other method in this class.

File: sdk/expo/test/helpers/TestmailInbox.ts (hunk 12 above)
Comment (vitalyrudenko-affinidi) [reply sdk/expo/test/helpers/TestmailInbox.ts:1]: I also don't like copy-pasting, but I'm not sure where to put it. I'd rather have some lib which has useful utilities, including tests -- and would put it there.
And it it referenced in `expo`, here:
https://github.com/affinityproject/affinidi-core-sdk/blob/cba04d447eb3720320f29b89ac39c478dff58d41/sdk/expo/test/integration/otp/index.test.ts#L12

File: sdk/browser/test/helpers/TestmailInbox.ts (hunk 11 above)
Comment (vitalyrudenko-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:34]: We don't need it cryptographically secure random, I just didn't know what else to use in this case.
Simplest solution for me was to use this lib.

One minute is arbitrary. Just in case there will ever be a collision, avoid situation where `livequery` immediately resolves with a very old email.

File: sdk/browser/test/helpers/TestmailInbox.ts (hunk 13 above)
Comment (vitalyrudenko-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:22]: We need to avoid using this helper just for email generation, unless we want to receive emails and read them.
It's because `testmail` is a paid product, and allows no more than 10,000 mails per month for our current plan.
So for random email generation (and we don't care about the mails) we can use `generateEmail()`.
If you understood me, it would be nice if you helped me to shape this comment into a useful state.

File: sdk/browser/test/integration/otp/index.test.ts (hunk 14 above)
Comment (vitalyrudenko-affinidi) [reply sdk/browser/test/integration/otp/index.test.ts:14]: It fails and just didn't want to mess more than I already did :D
Says "it's not a module" or something like that.

File: sdk/core/test/unit/services/CognitoService.test.ts (hunk 16 above)
Comment (vitalyrudenko-affinidi) [reply sdk/core/test/unit/services/CognitoService.test.ts:717]: No, because it restores in `afterEach`:
https://github.com/affinityproject/affinidi-core-sdk/blob/65ed77a4e28177ae727fe638be9932073bb67f18/sdk/core/test/unit/services/CognitoService.test.ts#L54

File: sdk/browser/test/helpers/TestmailInbox.ts (hunk 11 above)
Comment (inga-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:34]: But there should be no collision. And if we think there might be, it's better to increase the `inboxId` length to avoid them.

File: sdk/expo/test/helpers/TestmailInbox.ts (hunk 12 above)
Comment (inga-affinidi) [reply sdk/expo/test/helpers/TestmailInbox.ts:1]: We might just create some tiny package, like `@affinidi/testmail-client`

File: sdk/browser/test/helpers/TestmailInbox.ts (hunk 11 above)
Comment (vitalyrudenko-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:34]: Are there cons of doing it the way I did here?

File: sdk/browser/test/helpers/TestmailInbox.ts (hunk 13 above)
Comment (inga-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:22]: I think I start to understand it now.

But if no mail messages are sent, then it does not affect the remaining quota.

So we should only avoid using TestmailInbox in cases when:
* We need an email address; and
* Some messages will be sent to this address; and
* We don't care about these messages and will never read them.

Right?

I'm not sure it is such a common scenario for us to guard against it.

File: sdk/expo/test/helpers/TestmailInbox.ts (hunk 12 above)
Comment (vitalyrudenko-affinidi) [reply sdk/expo/test/helpers/TestmailInbox.ts:1]: Perhaps it can be done as a separate "polishing" task? Because I'm not sure if it's that much of a problem right now (not sure if we gonna change this code any time soon) and I'd rather focus on more important tasks.
@DenisPopov15 what do you think?

File: sdk/browser/test/integration/otp/index.test.ts (hunk 14 above)
Comment (inga-affinidi) [reply sdk/browser/test/integration/otp/index.test.ts:14]: Ah, of course, because `signedCredentials` is indeed not a modern module for some reason: https://github.com/affinityproject/affinidi-core-sdk/blob/master/sdk/browser/test/factory/signedCredentials.ts#L67

OK, let's leave it like that in this PR

File: sdk/browser/test/helpers/TestmailInbox.ts (hunk 13 above)
Comment (vitalyrudenko-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:22]: If I'm not mistaken, email is always sent in `signIn`/`signUp` methods, that are also tested in `NetworkMember.test.ts`. I haven't gone deep to actually test that, but I just wanted to point out that I'm discouraging usage if this helper (and testmail in general) for mere email generation. It doesn't make sense either way, since this service should only be used for testing emails.

File: sdk/browser/test/helpers/TestmailInbox.ts (hunk 11 above)
Comment (inga-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:34]: It introduces a magic constant and makes intent harder to follow, ihmo.
But if you feel it's better that way, then let's leave it that way.

File: sdk/core/test/unit/services/CognitoService.test.ts (hunk 16 above)
Comment (inga-affinidi) [reply sdk/core/test/unit/services/CognitoService.test.ts:717]: Oh. That's really hard to follow, because `afterEach` is on another level.
But since it was there before, let's leave it as it was.

File: sdk/expo/test/helpers/TestmailInbox.ts (hunk 12 above)
Comment (inga-affinidi) [reply sdk/expo/test/helpers/TestmailInbox.ts:1]: > And it it referenced in expo, here:

GitHub decided that I don't need to be able to find it 🙃

>  sdk/expo/test/integration/otp/index.test.ts 
> Load diff
> Large diffs are not rendered by default.

Thank you for the tip

Comment (inga-affinidi) [reply sdk/expo/test/helpers/TestmailInbox.ts:1]: > Because I'm not sure if it's that much of a problem right now (not sure if we gonna change this code any time soon)

AFAIK we're going to migrate our other projects to use testmail for testing, so it would be really useful to extract it into a separate package.
(And to not have three copies of the same file in affinidi-core-sdk.)

But priorities are up to @DenisPopov15

Comment (vitalyrudenko-affinidi) [reply sdk/expo/test/helpers/TestmailInbox.ts:1]: image

Comment (vitalyrudenko-affinidi) [reply sdk/expo/test/helpers/TestmailInbox.ts:1]: > > Because I'm not sure if it's that much of a problem right now (not sure if we gonna change this code any time soon)
> 
> AFAIK we're going to migrate our other projects to use testmail for testing, so it would be really useful to extract it into a separate package.
> (And to not have three copies of the same file in affinidi-core-sdk.)
> 
> But priorities are up to @DenisPopov15

Oh, then it would be useful to extract this, indeed. I need guidance, though, in were to put it and how.
Separate small package (`@affinidi/testmail-client`) that you suggested sounds good to me.

File: sdk/browser/test/helpers/TestmailInbox.ts (hunk 13 above)
Comment (inga-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:22]: > If I'm not mistaken, email is always sent in signIn/signUp methods, that are also tested in NetworkMember.test.ts.

If a new user is signed up by email address, then amazon does indeed send the message. However, in that case we would probably also need to get the confirmation code in order to move forward.

What I'm trying to say is:
* If we need to get the message, we should use TestmailInbox;
* If we just need some email address and no messages are getting sent, we should not use TestmailInbox (just because it was created for an entirely different purpose, as it name reflects); but using it will do no harm;
* And the cases where we need an email address, and some messages are getting sent to it, but we don't really care about their content, are not that common, and even if we use TestmailInbox there, it will not do much harm (maybe it will increase total number of messages by 10% or something like that). Even then, it is probably easier to use TestmailInbox there from the standpoint of code uniformity; and anyway, where else do we get a temporary address? Sending some test messages to testmail and some other to gdwk.in does not seem to be a good idea.

Comment (vitalyrudenko-affinidi) [reply sdk/browser/test/helpers/TestmailInbox.ts:22]: Still, I want to separate "testmail" inboxes and some random email generation.
Even 10% is huge amount, I've tried to minimize emails sent as much as I could, so I'm sticking to that - it allows us to keep using the cheaper plan of testmail, and also I don't think we should use a whole service helper just for an email generation.

> Sending some test messages to testmail and some other to gdwk.in does not seem to be a good idea.

We can just use something like `example.com`, it's not a problem.

File: sdk/browser/test/integration/otp/index.test.ts [hunk 17]
```diff
@@ -39,174 +29,154 @@ const credentialShareRequestToken =
...
-describe('AffinityWallet (flows that require OTP)', () => {
-  it('Save Open Attestation credential and #deleteCredential scenario', async () => {
-    const cognitoUsername = generateEmail()
+const messageParameters: MessageParameters = {
+  message: `Your verification code is: {{CODE}}`,
+  subject: `Verification code`,
+}
 
-    const token = await AffinityWallet.signUp(cognitoUsername, cognitoPassword, options)
+const waitForOtpCode = async (inbox: __dangerous.TestmailInbox): Promise<string> => {
```
Comment (inga-affinidi) [inline sdk/browser/test/integration/otp/index.test.ts:39]: Why is it in __dangerous part of the main code rather than in tests?

@DenisPopov15 , can we move TestmailInbox out into a separate npm package, please?

Comment (vitalyrudenko-affinidi) [reply sdk/browser/test/integration/otp/index.test.ts:39]: It's an okay-ish start. At least it will be easier to move to a separate package, since there is no copy paste. For a separate package I think there must be a separate task.

File: sdk/browser/test/integration/otp/index.test.ts (hunk 15 above)
Comment (vitalyrudenko-affinidi) [reply sdk/browser/test/integration/otp/index.test.ts:51]: Created a ticket: https://replika.atlassian.net/browse/FTL-1563

**Review by Volland — APPROVED** [review]

**Review by DenisPopov15 — APPROVED** [review]
